File: packages/js/gateway/src/session-storage.ts
import { v4 as uuidv4 } from 'uuid';
import type { HandoffLogEntry, HandoffSession } from '@trikhub/manifest';

// ============================================================================
// Session Storage Interface
// ============================================================================

/**
 * Interface for session storage implementations.
 * Manages HandoffSession lifecycle for the gateway.
 */
export interface SessionStorage {
  /** Create a new handoff session for a trik */
  createSession(trikId: string): HandoffSession;
  /** Get an existing session by ID */
  getSession(sessionId: string): HandoffSession | null;
  /** Append a log entry to a session */
  appendLog(sessionId: string, entry: HandoffLogEntry): void;
  /** Close a session (marks it as ended) */
  closeSession(sessionId: string): void;
}

// ============================================================================
// In-Memory Implementation
// ============================================================================

/**
 * In-memory session storage.
 * Sessions are lost on process restart — suitable for development and testing.
 */
export class InMemorySessionStorage implements SessionStorage {
  private sessions = new Map<string, HandoffSession>();

  createSession(trikId: string): HandoffSession {
    const now = Date.now();
    const session: HandoffSession = {
      sessionId: uuidv4(),
      trikId,
      log: [],
      createdAt: now,
      lastActivityAt: now,
    };
    this.sessions.set(session.sessionId, session);
    return session;
  }

  getSession(sessionId: string): HandoffSession | null {
    return this.sessions.get(sessionId) ?? null;
  }

  appendLog(sessionId: string, entry: HandoffLogEntry): void {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error(`Session "${sessionId}" not found`);
    }
    session.log.push(entry);
    session.lastActivityAt = Date.now();
  }

  closeSession(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error(`Session "${sessionId}" not found`);
    }
    session.lastActivityAt = Date.now();
  }
}

File: packages/python/trikhub-gateway/src/trikhub/gateway/session_storage.py
"""
Session storage for handoff sessions.

Mirrors packages/js/gateway/src/session-storage.ts
"""

from __future__ import annotations

import time
import uuid
from typing import Protocol

from trikhub.manifest import HandoffLogEntry, HandoffSession


class SessionStorage(Protocol):
    """Interface for session storage implementations."""

    def create_session(self, trik_id: str) -> HandoffSession: ...
    def get_session(self, session_id: str) -> HandoffSession | None: ...
    def append_log(self, session_id: str, entry: HandoffLogEntry) -> None: ...
    def close_session(self, session_id: str) -> None: ...


class InMemorySessionStorage:
    """
    In-memory session storage.
    Sessions are lost on process restart — suitable for development and testing.
    """

    def __init__(self) -> None:
        self._sessions: dict[str, HandoffSession] = {}

    def create_session(self, trik_id: str) -> HandoffSession:
        now = int(time.time() * 1000)
        session = HandoffSession(
            sessionId=str(uuid.uuid4()),
            trikId=trik_id,
            log=[],
            createdAt=now,
            lastActivityAt=now,
        )
        self._sessions[session.sessionId] = session
        return session

    def get_session(self, session_id: str) -> HandoffSession | None:
        return self._sessions.get(session_id)

    def append_log(self, session_id: str, entry: HandoffLogEntry) -> None:
        session = self._sessions.get(session_id)
        if session is None:
            raise ValueError(f'Session "{session_id}" not found')
        session.log.append(entry)
        session.lastActivityAt = int(time.time() * 1000)

    def close_session(self, session_id: str) -> None:
        session = self._sessions.get(session_id)
        if session is None:
            raise ValueError(f'Session "{session_id}" not found')
        session.lastActivityAt = int(time.time() * 1000)
